### PR TITLE
v12 SRF (per-chunk PDC support, no double-compression); fix bad bug causing chunks go poof.

### PR DESCRIPTION
**Pull request by kyngs**
This PR attempts to add per-chunk PDC support.
To test this functionality, one can use the plugin I've coded for this https://github.com/kyngs/ASP-PDCTest/tree/pr90

@ComputerNerd100 raised concerns about how I wrap NMSChunk around the original SlimeChunk, so please review that.
This is a bigger PR, so there may be some bugs, I've done a lot of testing tho.

*Hopefully I didn't forget anything.*

### CODE REVIEW

**Comment by MenschenToaster** [comment]
Could Issue #5 be implemented with the custom chunk compound tags as well? 🤔

**Comment by FabulousCodingFox** [comment]
In the current implementation, your patch 0011 clashes with the already existing patch 0011-Make-SlimeWorld-a-PersistentDataHolder in the 1.20.4 version

**Comment by kyngs** [comment]
> In the current implementation, your patch 0011 clashes with the already existing patch 0011-Make-SlimeWorld-a-PersistentDataHolder in the 1.20.4 version

hm, that one ain't from 1.20.4, but from main. Gonna do a quick rebase!

**Comment by kyngs** [comment]
Done, thanks for the tip

File: core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+            var entities = read(chunkData);
+
+            CompoundTag entitiesCompound = readCompound(entities);
+            @SuppressWarnings("unchecked")
+            List<CompoundTag> serializedEntities = ((ListTag<CompoundTag>) entitiesCompound.getValue().get("entities")).getValue();
+
+            // Extra Tag
+            var rawExtra = read(chunkData);
+            var extra = readCompound(rawExtra);
```
Comment (Paul19988) [inline core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java:133]: Remove usage of var

File: core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            // If the extra tag is empty, the serializer will save it as null.
+            // So if we deserialize a null extra tag, we will assume it was empty.
+            if (extra == null) extra = new CompoundTag("", new CompoundMap());
```
Comment (Paul19988) [inline core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java:136]: Can you wrap this in brackets for readability

File: core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    private static byte[] read(DataInputStream stream) throws IOException {
+        int length = stream.readInt();
+        byte[] data = new byte[length];
+        stream.read(data);
+        return data;
+    }
+
+    private static CompoundTag readCompound(byte[] tagBytes) throws IOException {
+        if (tagBytes.length == 0) return null;
```
Comment (Paul19988) [inline core/src/main/java/com/infernalsuite/aswm/serialization/slime/reader/impl/v12/v12SlimeWorldDeSerializer.java:162]: Can you wrap this in brackets for readability

File: patches/server/0012-Add-v12-chunk-pdc-and-extra-nbt.-Fix-double-compress.patch [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+diff --git a/src/main/java/com/infernalsuite/aswm/level/SlimeLevelInstance.java b/src/main/java/com/infernalsuite/aswm/level/SlimeLevelInstance.java
+index 65b475b1292e01c918c1f8144599b5fa78688e97..67be177f92ff66074b6f0f0f82fce284ccb229a4 100644
+--- a/src/main/java/com/infernalsuite/aswm/level/SlimeLevelInstance.java
++++ b/src/main/java/com/infernalsuite/aswm/level/SlimeLevelInstance.java
+@@ -151,9 +151,9 @@ public class SlimeLevelInstance extends ServerLevel {
+             Bukkit.getLogger().log(Level.INFO, "Saving world " + this.slimeInstance.getName() + "...");
+             long start = System.currentTimeMillis();
+ 
+-            Bukkit.getLogger().log(Level.INFO, "CONVERTING NMS -> SKELETON");
++            //Bukkit.getLogger().log(Level.INFO, "CONVERTING NMS -> SKELETON");
```
Comment (Paul19988) [inline patches/server/0012-Add-v12-chunk-pdc-and-extra-nbt.-Fix-double-compress.patch:183]: If we're going to comment this out, we should remove it

File: patches/server/0012-Add-v12-chunk-pdc-and-extra-nbt.-Fix-double-compress.patch [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+             SlimeWorld world = this.slimeInstance.getForSerialization();
+-            Bukkit.getLogger().log(Level.INFO, "CONVERTED TO SKELETON, PUSHING OFF-THREAD");
++            //Bukkit.getLogger().log(Level.INFO, "CONVERTED TO SKELETON, PUSHING OFF-THREAD");
```
Comment (Paul19988) [inline patches/server/0012-Add-v12-chunk-pdc-and-extra-nbt.-Fix-double-compress.patch:186]: If we're going to comment this out, we should remove it

**Review by Paul19988 — CHANGES_REQUESTED** [review]

**Review by Paul19988 — APPROVED** [review]

**Comment by ComputerNerd100** [comment]
@Paul19988 @kyngs Ty for the var changes
Almost happy with this my end too
I'll update this to target develop shortly and then merge to there and we can get some snapshot releases out

**Review by ComputerNerd100 — APPROVED** [review]

**Pull request merged**